Adding Multinomial and Nested Logit Models for Consumer Choice

NathanielF opened the pull request:
## Description
I'm adding two new model classes for discrete choice style models that I intend to be part of the consumer choice module. 
As it stands i'm opening the PR as a draft for discussion around the implementation choices and API design I have for these models. 

Related to this issue: https://github.com/pymc-labs/pymc-marketing/issues/1653. There is a lot of potential in the discrete choice style models for Bayesian modelling in particular, the state of the art models in this domain involves a mixed logit parameterisation for which "vanilla" implementations are pretty straightforward using Bayesian hierarchical parameterisations. 

### Two New Models

Main things to flag: There are now two new model files in the consumer choice folder. The simple Multinomial Logit and the Nested Logit. As outlined in the issue i've restricted the nested logit to no more than two layers of nesting. I believe this will bring us to beyond parity with packages like `mlogit` in R and `pylogit` in python which allow for only 1 level deep nesting structures. 

### API Discussion

The API i'm suggesting for these models differs from the typical X,y inputs on the models in pymc marketing in general. Mostly this is because I feel the use of Wilkinson style notation here is important. For instance this is how you specify the Nested Logit Model currently: 

<img width="1048" alt="image" src="https://github.com/user-attachments/assets/93e943d2-49bd-415a-9c7f-07e8ccab5042" />

We assume a wide-data input as well: 

<img width="952" alt="image" src="https://github.com/user-attachments/assets/8d8984cf-8755-4f9f-ba6b-ddcb8b6b8668" />

### Causal Inference and Counterfactuals

The value that these models bring is their focus on causal inference. The entire history of discrete choice models stems effectively from the observation that multinomial logit models cannot support plausible counterfactuals around market interventions (due to IIA) and more sophisticated discrete choice models like the nested logit models are able to solve this. See for instance here how a pricing intervention on a multinomial logit results in proportional re-allocation of market share to the rest of the market.  

<img width="941" alt="image" src="https://github.com/user-attachments/assets/6f81191c-d781-411a-a007-e6d2eac0302a" />

We demonstrate this problem and solution by adding 2 new notebooks to the gallery. 

<img width="770" alt="image" src="https://github.com/user-attachments/assets/a3afc256-6db0-4e9e-817b-dc5c3e21f277" />

In the second notebook for nested logit we show how the IIA is solved by this extra nesting structure: 

<img width="928" alt="image" src="https://github.com/user-attachments/assets/538b9899-f2b7-4c14-b81a-a47312075705" />

### Fixed Attributes and Alternative Specific Attributes

One thing i've done is to ensure that the models can identify parameters for the alternative specific attributes (e.g. price) and the individually fixed attributes e.g. (income). I've done my best to benchmark the parameter identification and recovery against R's `mlogit` package: 



### How to Proceed?

I have not done an extensive write up of the math behind these types of models and some of the functions need more documentation and tests. But I wanted to share what I have so far to generate discussion and maybe decide on how to proceed. One immediate improvement i could think of would be to remove duplication from the nested logit and multinomial logit model classes, making them instances of a more general "DISCRETE CHOICE" class where we could re-use e.g. the formula parsing functions. Additionally i'd like to benchmark the parameter identification with a second data set and example. 

Longer term i think there is room for adding a vanilla mixed-logit example too. 

Anyway, open to feedback. Adding a draft PR now to check which linting, and testing failures i have. 

## Related Issue
- [ ] Closes #
- [x] Related to  https://github.com/pymc-labs/pymc-marketing/issues/1653

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://www.pymc-marketing.io/en/latest/contributing/index.html). Feel free to comment [`pre-commit.ci autofix` to auto-fix](https://pre-commit.ci/#configuration-autofix_prs).
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks) using [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--1654.org.readthedocs.build/en/1654/

<!-- readthedocs-preview pymc-marketing end -->

williambdean commented:
Can this be done with bambi?

NathanielF commented:
Some aspects of the multinomial logit I believe so, but not the separate utility equations with fixed covariates and the nested logit cannot be done with Bambi.

NathanielF commented:
Maybe to make that a little clearer @williambdean the Multinomial logit in this discrete choice implementation is related to the more standard multinomial regression you will find in Bambi, but it differs is an important way that is rooted in the utility theory behind the modelling enterprise. The model is conceptualised as involving "drivers" of the utility for each of the products on a market, so within the model with N linear models which represent the utility of that good where each of the models takes attributes of the product alternative as features. By allowing these distinct "alternative specific" and "individual specific" covariates we attempt to model a choice-scenario and the covariates have a specific interpretation under scenario. 

Standard multinomial regression models don't make that distinction and so can't be interpreted in the same way. 

Do you think it would help the PR if i put more of this background in the notebooks?

NathanielF commented:
If it helps @juanitorduz the nested logit implementation comes from here: https://eml.berkeley.edu/books/choice2nd/Ch04_p76-96.pdf

I can add more write up on the model itself if it helps the PR later. But would be useful even just to have high level feedback on the  API and function call patterns etc...

NathanielF commented on pymc_marketing/customer_choice/mnl_logit.py:
This is multinomial and yes we could definitely spell it out.

NathanielF commented on pymc_marketing/customer_choice/mnl_logit.py:
```diff
@@ -0,0 +1,586 @@
+                    pt.set_subtensor(betas_fixed_[-1, :], 0),
+                    dims=("alts", "fixed_covariates"),
+                )
+                F_data = pm.Data("F", F)
```
alts = alternatives refers to the different products being considered i.e. one for each of the utility equations. 

fixed_covariates refers to the number of covariates in each equation that do not vary across the equations e.g. your income is the same when you consider choices between bus, car and plane... but the price of bus, car and plane vary across the goods. To identify the parameters for the fixed covariates we need to either interact them with the intercept (as is done in mlogit and traditional frequentist discrete choice models) or more flexibly in the Bayesian setting allow a parameter on the fixed covariate for each of the goods. 

In the API as seen in the notebooks: 

```
new_utility_formulas = [
    "gc ~ ic_gc + oc_gc | income + rooms + agehed",
    "gr ~ ic_gr + oc_gr | income + rooms + agehed",
    "ec ~ ic_ec + oc_ec | income + rooms + agehed",
    "er ~ ic_er + oc_er | income + rooms + agehed",
    #'hp ~ ic_hp + oc_hp | income + rooms + agehed'
]

```

I put the fixed covariates behind the bar `|` and so the dimension of `fixed_covariates` are dependent on how many the user puts into the model

williambdean replied:
Can we name the dimensions and add to the model?

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+
+class NestedLogit(ModelBuilder):
+    """
+    Nested Logit class.
```
Could we add some references to the doctrings? (books or papers?)

NathanielF replied:
Done

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+
+        return result
+
+    def parse_formula(self, df, formula, depvar):
```
type hints missing

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+
+        return target, alt_covariates, fixed_covariates
+
+    def prepare_X_matrix(self, df, utility_formulas, depvar):
```
type hints

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        """
+        n_obs = len(df)
+        n_alts = len(utility_formulas)
+        n_covariates = len(utility_formulas[0].split("|")[0].split("+"))
```
maybe this can be encoded in a function = len(_get_covariates(...)). This way we can unit-test `_get_covariates`?

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        return X, F, alts, np.unique(fixed_covariates)
+
+    @staticmethod
+    def _prepare_y_outcome(df, alternatives, depvar):
```
type hints missing

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        return y, prod_mapping
+
+    @staticmethod
+    def _parse_nesting(nest_dict, product_indices):
```
type hints missing

NathanielF replied:
added

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        return coords
+
+    def preprocess_model_data(
+        self, choice_df, utility_equations
```
type hints missing

juanitorduz replied:
alsmo missing parameter description in docstrings

NathanielF replied:
added

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        """Do not use, required by parent class. Prefer make_model()."""
+        return super().build_model(X, y, **kwargs)
+
+    def make_exp_nest(self, U, W, betas_fixed_, lambdas_nests, nest, level="top"):
```
type hints missing

juanitorduz replied:
Also add parameter description in the doctrings

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        return super().build_model(X, y, **kwargs)
+
+    def make_exp_nest(self, U, W, betas_fixed_, lambdas_nests, nest, level="top"):
+        """Calculate within nest probabilities.
```
Can you explain a bit more the logic of this in the doctrings?

NathanielF replied:
done

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        exp_W_nest = pm.math.exp(W_nest)
+        return exp_W_nest, P_y_given_nest
+
+    def make_P_nest(self, U, W, betas_fixed_, lambdas_nests, level):
```
type hints missing and parameter descriptions

juanitorduz replied:
maybe even add some comments on the code to explain the logic a bit? (It is not super easy to follow for a new person reading for the first time)

NathanielF replied:
added

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+            path_prods_t = []
+            ordered = [(key, min(vals)) for key, vals in nest_indices["top"].items()]
+            top_nests = [x[0] for x in sorted(ordered, key=lambda x: x[1])]
+            for _idx, n in enumerate(top_nests):
```
An empty line before this line to make the code more readable?

NathanielF replied:
done

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+
+        return attrs
+
+    def sample_prior_predictive(self, extend_idata, kwargs):
```
type hints missing

NathanielF replied:
added but i had to override parent class function signature because mypy was complaining

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+            with self.model:
+                self.idata = pm.sample(**kwargs)
+
+    def sample_posterior_predictive(self, extend_idata, kwargs):
```
type hints missing

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        )
+        return self
+
+    def apply_intervention(self, new_choice_df, new_utility_equations=None):
```
type hints missing and parameters in docstings

NathanielF replied:
added

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        return idata_new_policy
+
+    @staticmethod
+    def calculate_share_change(idata, new_idata):
```
type hints missing and parameters in docstings

NathanielF replied:
Done

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+            shares_df_new, left_on="product", right_on="product", how="left"
+        )
+        shares_df.fillna(0, inplace=True)
+        shares_df["relative_change"] = (
+            shares_df["new_policy_share"] - shares_df["policy_share"]
+        ) / shares_df["policy_share"]
+        shares_df.set_index("product", inplace=True)
```
What about using https://pandas.pydata.org/docs/reference/api/pandas.eval.html to make these computations easier to read?

NathanielF replied:
Tried this, just FYI it works nicely in code but breaks in the testing suite for 3.10

<img width="1030" alt="image" src="https://github.com/user-attachments/assets/0ce0dde0-16e0-45d7-a73c-ea643081c44e" />

juanitorduz commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,783 @@
+        return shares_df
+
+    @staticmethod
+    def make_change_plot(change_df, title="Price Intervention", figsize=(8, 4)):
```
type hints missing and parameters in docstings

juanitorduz reviewed (changes requested):
Thanks @NathanielF ! This looks very interesting. I do not have much experience with these models but they make sense at first glance ;) My comments are moe on the form of code. We need to make sure we have complete type hints (input and output), parameter description in the docstrings and ideally references on the formulas we are using (either from books, papers or other online examples). 

I think once these style requirements are done, we can merge this one and release as an experimental feature/

williambdean commented on pymc_marketing/customer_choice/mnl_logit.py:
```diff
@@ -0,0 +1,586 @@
+        betas_fixed = Prior("Normal", mu=0, sigma=1, dims=("alts", "fixed_covariates"))
+
+        return {
+            "alphas_": alphas,
+            "betas": betas,
+            "betas_fixed_": betas_fixed,
```
Why do some have suffix of _ and others do not?

NathanielF replied:
I need to 0 out one of the alternatives alphas_ --> alpha and I was doing the same for betas_fixed_ for parameter identification. 

e.g alpha = pm.Deterministic('alpha', pt.set_subtensor(alphas_, -1, 0))....

williambdean commented on pymc_marketing/customer_choice/mnl_logit.py:
```diff
@@ -0,0 +1,586 @@
+
+        return result
+
+    def parse_formula(self, df, formula, depvar):
```
Can we pull out this into a function instead of method. A good separation might be: 
1. parse_formula(formula) -> tuple[target, alt_covariates, fixed_covariates]
2. check_columns(covariates, df)
3. check_dependent_variable(target, ser)

NathanielF replied:
Yeah, can do... probably not tonight though. Wrecked. Will try and pick it up during the week.

NathanielF commented:
Ok folks. Making progress. 

Added a bunch of type-hints and improved doc-strings to the nested logit class. Also added a light write up to both notebooks for multinomial and nested logit on the whys and hows. 

Still missing a few doc-strings and will address @williambdean 's request soon too. 

Thanks for the feedback @juanitorduz , @williambdean .

williambdean commented on pymc_marketing/customer_choice/mnl_logit.py:
```diff
@@ -0,0 +1,586 @@
+        }
+        return coords
+
+    def preprocess_model_data(self, choice_df, utility_equations):
```
Thanks for explaining. Would it make sense to have a default of None which would use the self.utility_formulas?

williambdean commented:
Really cool stuff. I liked those videos that were linked

williambdean commented on pymc_marketing/customer_choice/mnl_logit.py:
```diff
@@ -0,0 +1,586 @@
+        """Do not use, required by parent class. Prefer make_model()."""
+        return super().build_model(X, y, **kwargs)
+
+    def make_model(self, X, F, y) -> None:
```
Is it possible to put X and F in a single xr.Dataset and pass as "X"

Or would that be unituitive?

NathanielF replied:
I think that wouldn't be in the spirit of the model, the decomposition to fixed covariates and alternative specific covariates is pretty foundational to the literature and write up. Would avoid bundling them if i could.

williambdean commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,1064 @@
+        coords = self.coords
+
+        with pm.Model(coords=coords) as model:
+            alphas = pm.Normal("alphas", 0, 1, dims="alts")
+            betas = pm.Normal("betas", 0, 1, dims=("alt_covariates"))
```
Are these priors to expose in the configuration?

NathanielF replied:
Good catch, they weren't being passed through. I've amended that now. 
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/c2a56d2f-e502-4ae9-a903-9ef133f55c96" />

williambdean commented on pymc_marketing/customer_choice/nested_logit.py:
```diff
@@ -0,0 +1,1064 @@
+        self, extend_idata: bool, kwargs: dict[str, Any]
+    ) -> None:
+        """Sample Posterior Predictive Distribution."""
+        if extend_idata:
+            with self.model:
+                if self.idata is not None:
+                    self.idata.extend(
+                        pm.sample_posterior_predictive(
+                            self.idata, var_names=["likelihood", "p"], **kwargs
+                        )
+                    )
+        else:
+            with self.model:
+                self.post_pred = pm.sample_posterior_predictive(
+                    self.idata, var_names=["likelihood", "p"], **kwargs
+                )
```
nit: Can we sample_posterior_predictive then conditionally extend or assign instead of duplicating much of the logic

```python
with self.model: 
    idata = pm.sample_posterior_predictive(...)

if extend_idata ...
```

NathanielF commented on docs/source/notebooks/customer_choice/mnl_logit.ipynb:
```diff
@@ -0,0 +1,1362 @@
+{
```
Done.

NathanielF replied:
<p>Done</p>

NathanielF commented on docs/source/notebooks/customer_choice/mnl_logit.ipynb:
```diff
@@ -0,0 +1,1362 @@
+{
```
<p><a href="https://github.com/pymc-labs/pymc-marketing/blob/b13b67a0d517661464b93c2a17d965377d13c4fd/pymc_marketing/customer_choice/nested_logit.py#L224" target="_blank">https://github.com/pymc-labs/pymc-marketing/blob/b13b67a0d517661464b93c2a17d965377d13c4fd/pymc_marketing/customer_choice/nested_logit.py#L224</a></p><p> </p><p>This should be checked here</p>

NathanielF commented on docs/source/notebooks/customer_choice/nested_logit.ipynb:
```diff
@@ -0,0 +1,2488 @@
+{
```
<p>No, these should be handled here: <a href="https://github.com/pymc-labs/pymc-marketing/blob/b13b67a0d517661464b93c2a17d965377d13c4fd/pymc_marketing/customer_choice/nested_logit.py#L224" target="_blank">https://github.com/pymc-labs/pymc-marketing/blob/b13b67a0d517661464b93c2a17d965377d13c4fd/pymc_marketing/customer_choice/nested_logit.py#L224</a></p>

NathanielF commented on pymc_marketing/customer_choice/mnl_logit.py:
```diff
@@ -0,0 +1,586 @@
+        return shares_df
+
+    @staticmethod
+    def make_change_plot(change_df, title="Price Intervention", figsize=(8, 4)):
```
changed to "plot_..."

juanitorduz commented:
Thanks @NathanielF ! Once @williambdean gives the last ✅ we can merge and release to get feedback from the community! 🤘

NathanielF commented:
Exciting @juanitorduz ! Real curious to see if these models are useful for the community. There is quite a few directions we could go with them.

williambdean commented on docs/source/notebooks/customer_choice/nested_logit.ipynb:
```diff
@@ -0,0 +1,2488 @@
+{
```
<p>Alright. Then let's leave out for now</p>

juanitorduz commented:
Actually, this PR looks great, and I am sure we can improve in future iterations. So let's merge and iterate :)
